### DISCUSSION TITLE
Serial Device: RX buffer termination vs. timeout

### DISCUSSION THREAD

**Discussion by HoWeBrz** [pymeasure/pymeasure#633] [Q&A]
As discussed in #631 I want to receive serial data. 
In a "raw/pure" serial connection I am receiving the data properly after the RX timeout has been occured.
I simply need to readout the RX buffer via 
```
ser = serial.Serial()  			# get port handle
ser.port = 'COM6'				
ser.baudrate = 9600
ser.timeout =3					#read timeout
ser.open()						# open port
ser.is_open						# true if open	

ser.write(b"##JT:reset\n")
ser.write(b"##JI:0\n")
ser.write(b"##JD2:2\n")
ser.write(b"##JT:send\n")

time.sleep(1)
rxdat = str(ser.read(1000), 'utf-8')          #read rx buffer
print("Received: " + rxdat)
```

When doing this with an instrument via
```
rxdat = fb1.adapter.connection.read_bytes(1000)
print("Received: " + rxdat)
```
it is working if I do not attempt to read more bytes as in the buffer available.
But I do not know how much bytes will be in the buffer, so I have to wait some time (e.g. 1s) and then read the buffer.
But then the timeout occurs and throws an error.

How can I create my instrument such that I can:
1. Clear the RX buffer initially at startup ?
2. Wait some time before reading the RX buffer ?
3. Read the RX buffer without an timeout-error and without any termination character ?
4. What is the difference between "fb1.read()" and " fb1.adapter.connection.read_bytes(1000)" with regard to timing and termination ?

Here is my current instrument implementation:

```
from pymeasure.adapters.serial import SerialAdapter
from pymeasure.instruments import Instrument
from pymeasure.instruments.validators import truncated_range, strict_discrete_set, strict_range
from pyvisa.errors import VisaIOError
import time

class FuseBoard(Instrument):
    """Represents a FuseBoard device"""

    def __init__(self, resource_name, baud_rate=115200, **kwargs):
        super(FuseBoard, self).__init__(
            resource_name,
            "FuseBoard",
            baud_rate=baud_rate,
            data_bits=8,
            stop_bits=10,
            parity=0,
            timeout=1000,
            read_termination='\n',
            write_termination='\n',
            **kwargs
        )
    
    def write(self, command):
         """Override the existing write method from adapter."""
         print("My command:" + command)
         self.adapter.connection.write(command)

#############################################
# main function startes here
if __name__ == "__main__":
    
    # init FuseBoard device
    fb1 = FuseBoard("ASRL8::INSTR")
   
    # write serial commandos
    fb1.write("##JT:reset")    
    fb1.write("##JI:0")
    fb1.write("##JD2:2")
    fb1.write("##JT:send")

    #read RX buffer
    time.sleep(3)
    rxdat1 = fb1.adapter.connection.read_bytes(1000)
    print(rxdat1)
 
    #close connection
    fb1.adapter.connection.close()
```

**Comment by HoWeBrz** [comment]
Question 4 was already answered in the referenced Disussion #631. I missed that before - Thx for your explanation.

**Comment by bilderbuchi** [comment]
If this 
```
    # write serial commandos
    fb1.write("##JT:reset")    
    fb1.write("##JI:0")
    fb1.write("##JD2:2")
    fb1.write("##JT:send")

    #read RX buffer
    time.sleep(3)
    rxdat1 = fb1.adapter.connection.read_bytes(1000)
    print(rxdat1)
```
is one "thing" you do (i.e. always those commands), I would put this into a method `get_JTAG_status()` or so. Instead of printing, return rxdat1.

re 1: you can read single bytes in a loop with a short timeout until nothing is coming back.
re 2: a sleep is fine in the method.
re 3: If you don't know the length of what's coming back (so you can't stop after correct nr of bytes), you have to wait for the timeout. Put the read_bytes call in a try-except, except the TimeoutError that you get so far and just put a `pass` in the clause - you handled the error.

**Reply by HoWeBrz** [reply]
I tried option 3 like this:
```
   fb1.write("##JT:reset")    
    fb1.write("##JI:0")
    fb1.write("##JD2:2")
     fb1.write("##JT:send")
    time.sleep(3)
    try:
        rxdat1 = fb1.adapter.connection.read_bytes(1000)
    except:
        pass    
    print(rxdat1)
```
but it tells me `NameError: name 'rxdat1' is not defined`

How can I get the read-bytes back from the "try:" condition ?

**Reply by HoWeBrz** [reply]
This somehow works - reading one byte in a loop until the timeput occurs: 
```
    fb1.write("##JT:reset")    
    fb1.write("##JI:0")
    fb1.write("##JD2:2")
    fb1.write("##JT:send")


    time.sleep(3)
    
    rxdat1 = ""    #empty string to collect single bits
    
    while(True):    
        try:
            rxdat1 += str(fb1.adapter.connection.read_bytes(1),'utf-8')    #read one byte in a loop
        except:
            break
    
    print(rxdat1)
```
Is this the proposed way to use the timeout + exception ?

My preferred way to read serial data would be to use "fb1.read()" with dedicated termination characters !

**Reply by bilderbuchi** [reply]
> but it tells me NameError: name 'rxdat1' is not defined

Ah, a complication - the timeout makes the call except, and thus rxdat1 never gets assigned, I did not think about that. Yes, accumulating byte by byte as a practicable way. Possibly that is what serial is doing internally, too.

> My preferred way to read serial data would be to use "fb1.read()" with dedicated termination characters !

This is everybody's preferred way! However, unfortunately your device does not grace us with proper message termination characters. :-/
You can try to use fb1.read() in a loop (I think \r\n is a general line termination character for your device) until no more lines appear, and assemble that into a message.

**Comment by HoWeBrz** [comment]
What I understood is, that the "default" termonatino character "\n" or "\r" will not work, because the serial-response contains multiple "...\r\r\n....\n...".  Thus we need another termination charactor/sequence/indicator.

We agree on a dedicated end-of-transmission indicator "##EOT", but apparently this does not work as expected.
The response - captured by the pure serial connection looks like this:
` b'##JT:reset\r\n##JI:0\r\n##JD2:2\r\n##JD98:8022\r\n##JD100:10062\r\n##JD109:160000\r\n##JT:send\r\n\r\r\n\r\r\nJTAG Transmission... \r\r\n-------\r\r\n##JP0:\r\r\n##JR2:0\r\r\n##JR98:fffffffffffffffff00000003\r\r\n##JR100:fffffffffffffffff00000003\r\r\n##JR109:ffffffffffffffffffff0000000f\r\r\n##EOT\r\r\n'`


Applying the read termination characters
```
def __init__(self, resource_name, baud_rate=115200, **kwargs):
        super(FuseBoard, self).__init__(
            resource_name,
            "FuseBoard",
            baud_rate=baud_rate,
            data_bits=8,
            stop_bits=10,
            parity=0,
            timeout=5000,
            read_termination='##EOT\r\r\n',
            write_termination='\n',
            **kwargs
        )
```
or variations of it does not help for a successful

```
 try:
        rxdat = fb1.read()
 except:
        pass
print(rxdat)
```
instead it throws a

```
UserWarning: read string doesn't end with termination characters
  return self.connection.read()
```

How can I defined a proper read-termination to be able to use "fb1.read()"

**Reply by bilderbuchi** [reply]
Hm, this should work. [Here](https://pyvisa.readthedocs.io/en/latest/introduction/resources.html?highlight=termination#termination-characters) is the pyvisa documentation on read_termination, basically this is where we're forwarding the kwarg to.
Does it work with pure Pyvisa (like [this](https://pyvisa.readthedocs.io/en/latest/introduction/communication.html#communicating-with-your-instrument))?

**Reply by bilderbuchi** [reply]
> UserWarning: read string doesn't end with termination characters
  return self.connection.read()

OK, the reason is that PyVISA [actually](https://github.com/hgrecco/pyvisa/blob/5522e02a965bcd48a0beff2756ea402570b97db4/pyvisa/resources/messagebased.py#L398-L403) only uses the _last_ character of the termination sequence to detect when to stop reading, and then compares if there is a full match to the read_termination sequence. If not, you get that warning (as it stops reading at the first \n)

It seems you can influence the message format of this device -- can you make it use \r as the message/read_termination character and \n to separate lines in the messages itself?

**Reply by HoWeBrz** [reply]
Thanks for digging into this issue with pyVisa.

Currently I am using the "read-one-byte in a loop until timeout" workaround as explained above.
This is working - I think we will not change the message format right now.

**Reply by bilderbuchi** [reply]
Sure. I think in general it's probably useful to avoid "unconventional" message formats/terminators if you can. Maybe look to how other instruments tackle this - e.g. oscilloscope trace download or similar. A known message length for the JTAG dump would also help and avoid this whole conversation (I think). :-)
